src/stemmabench/algorithms/utils.py

```diff
@@ -0,0 +1,140 @@
+            return file.read()
+
+    @staticmethod   
+    def dict_of_children(edges: list[list[str]]) -> Dict[str, dict]:
+        """Builds a dictionary with node label as key and dictionary of children as value.
+
+        ### Args:
+            - edges (list): A list representation of edges in list format.
+
+        ### Returns:
+            - dict: A dictionary containing with a node as key and a dictionary children.
+
+        ### Example:
+            >>> Utils.dict_of_children([['a','b'], 
+                                        ['a','c'], 
+                                        ['b', 'd'], 
+                                        ['b','e'], 
+                                        ['c','f'], 
+                                        ['c', 'g']])
+            {'a': {'b': {}, 'c': {}}, 
+            'b': {'d': {}, 'e': {}}, 
+            'c': {'f': {}, 'g': {}}}
+        """
+        tree_data: dict = {}
+        for edge in edges:
+            if edge[0] not in tree_data:
+                tree_data[edge[0]] = {}
+            tree_data[edge[0]][edge[1]] = {}
+        return tree_data
+
+    @staticmethod
+    def find_root(dict_of_children: Dict[str, dict]) -> list[str]:
+        """Finds the root of the stemma from a dictionary of children.
+        
+        ### Args:
+            - dict_of_children (dict): Dictionary containing the labels of all manuscripts that have children as keys
+            and dictionary of whith all children of key manuscript as keys as values.
+        
+        ### Example:
+            >>> Utils.find_root({'a': {'b': {}, 'c': {}}, 
+                                 'b': {'d': {}, 'e': {}}, 
+                                 'c': {'f': {}, 'g': {}}})
+            ['a']
+        """
+        not_root = []
+        for i in dict_of_children:
+            for j in dict_of_children:
+                if str(i) in list(dict_of_children[j].keys()):
+                    not_root.append(i)
+        root = list(set(dict_of_children.keys()) - set(not_root))
+        return root
+
+    @staticmethod
+    def dict_from_edge(*,edge_path: Union[str, None] = None, edge_list: Union[list[list[str]], None] = None) -> Dict[str, dict]:
+        """Converts edge file to dictionary representation.
+
+        ### Args:
+            - edge_path (str, Optional): The path to the edge file used to construct dictionary.
+            - edge_list (str, Optional): The list of all the edges in a stemma tree.
+
+        ### Returns:
+            - dict: Dictionary representation of the tree in edge file.
+
+        ### Raises:
+            - ValueError: If edge file or list is not a valid edge (see method validate_edge for more info).
+            - ValueError: If no parameter is specified.
+            - ValueError: If both parameters are specified.
+        """
+        if edge_path and edge_list:
+            raise ValueError("Only one of the parameters edge_path and edge_list can be specified.")
```

Comment:
f-string possibilities here

Reply:
and actually for all errors raised at a later time

Reply:
Not sure f-strings would be relevent her as it is the names of the parameters that are being referenced